### PR TITLE
Support JPY, IDR and SGD currencies

### PR DESCRIPTION
**Pull request by guzzilar**
> This PR required PR #40 to be merged first.

#### 1. Objective

To make plugin support the following currencies.
- JPY (Japan)
- IDR (Indonesia)
- SGD (Singapore)

**Related information**:
Related issue(s): 🙅
Related ticket(s): T2244, T2245

#### 2. Description of change

- Add condition, if store's currency is not one of "THB", "JPY", "IDR", "SGD". Omise payment method will show warning message that the current currency is not support with Omise service.
 
- Be able to create charge with "THB", "JPY", "IDR", "SGD" currencies.
- Remove `Currency` configuration from the payment setting page, now using store's base currency configuration instead.
- Format the display of transaction amount at the plugin dashboard page.
- Remove placeholder that mentioned about `satang` from `create transfer input field` at the plugin dashboard page.
- Remove condition to hide `create transfer input field` if balance is lower than the minimum-transfer-amount. (this form used to support only `THB`, now I removed it out and let API says error if merchant cannot create transfer).

#### 3. Quality assurance

**🔧 Environments:**

- **Platform version**: Magento CE 1.9.3.1
- **Omise plugin version**: Omise-Magento 1.9.0.6
- **PHP version**: 5.6.28

**✏️ Details:**

1. ✅ If enter to the payment setting page (`System > Configuration > Sales > Payment Method`). "Currency" configuration is gone.
  screen shot 2560-01-24 at 2 56 08 pm

2. ✅ Test checkout with "JPY" currency
    2.1 Go to `System > Configuration > General > Currency Setup` to change base currency to `Japanese Yen`

    2.2. Set product's price (in this test, I set it to `20,000 YEN`) and do normal checkout

    screen shot 2560-01-24 at 5 27 32 pm

    2.3. Check result from Omise log, `currency` must be `JPY` and amount must be the same as the product price (0 decimal, no-subunit) (i.e. set product price to `15,025`, the amount that send to Omise must be `15025`)

    screen shot 2560-01-24 at 5 32 03 pm

3. ✅ Test checkout with "IDR" currency
    3.1. Go to `System > Configuration > General > Currency Setup` to change base currency to `Indonesian Rupiah`

    3.2. Set product's price (in this test, I set it to `Rp20,000`) and do normal checkout

    screen shot 2560-01-24 at 5 06 05 pm

    3.3. Check result from Omise log, `currency` must be `IDR` and amount must be subunit (i.e. set product price to `15,025`, the amount that send to Omise must be `1502500`).

    screen shot 2560-01-24 at 5 09 00 pm

4. ✅ Test checkout with "SGD" currency
    4.1 Go to `System > Configuration > General > Currency Setup` to change base currency to `Singapore Dollar`

    4.2. Set product's price (in this test, I set it to `20,000 SGD`) and do normal checkout
    screen shot 2560-01-24 at 5 56 21 pm

    4.3. Check result from Omise log, `currency` must be `SGD` and amount must be subunit (i.e. set product price to `15,025`, the amount that send to Omise must be `1502500`).

    _note, SGD currency is currently not supported, this test was made with Omise Japan account. But basically, just test that those parameters were passed to Omise API correctly._
    screen shot 2560-01-24 at 6 02 39 pm

5. ✅ Test set base currency to the currency that Omise doesn't support. Omise payment method shows warning message that the current currency is not support with Omise service.
    _Tested set base currency to `USD`_
    screen shot 2560-02-07 at 11 54 19 pm

6. ✅ Test if use currency that doesn't match with the currency of merchant account, it must raise an error. (i.e. this test set Magento's currency to `JPY` but use `Omise Indo account`)
    _note, error message came from Omise API_
    screen shot 2560-01-24 at 5 27 51 pm

7. ✅ At the plugin dashboard page, THB currency should display an amount as `฿12,000` instead of `12,000.00`.
    screen shot 2560-01-24 at 10 06 32 pm

8. ✅ At the plugin dashboard page, JPY currency should display an amount as `30,133円`.
    screen shot 2560-01-24 at 10 10 05 pm

9. ✅ At the plugin dashboard page, IDR currency should display an amount as `Rp54,000`.
    screen shot 2560-01-24 at 10 15 20 pm

10. ✅ At the plugin dashboard page, SGD currency should display an amount as `S$54,000`.

    _note, this test I used Omise Indo account to test, but bypass currency (hard-code to `SGD` instead of use variable) at `Omise_Gateway_Helper_Data` to see the result._
    _That is the reason I got these numbers._
    screen shot 2560-01-25 at 11 14 13 am

11. ✅ For changed that "- Remove condition to hide `create transfer input field` if balance is lower than the minimum-transfer-amount. (this form used to support only `THB`, now I removed it out and let API says error if merchant cannot create transfer)."

    Please check the screenshot below, at the transfer box. It'd show like 👇
    screen shot 2560-01-24 at 10 58 20 pm

#### 4. Impact of the change

Nothing.

#### 5. Priority of change

Normal.

#### 6. Additional Notes

- SGD currency is currently not supported from Omise API. This pull request just for prepare the plugin to be ready for coming feature.

- Reviewed by @nimid 
    ```
    The conclusion below is about the new support for a currency SGD which it has not been available at the Omise API side.

    When the currency SGD is available at the Omise API side, it still require the actual system test. Although, this pull request has provided the result of test on the 4th and 10th but the 4th is the result at the error log on Omise dashboard and the 10th is test by hard-code.
    For this situation, another test strategy such as unit test or another test can help.
    ```
    ref. https://github.com/omise/omise-magento/pull/44#issuecomment-278260551

### CODE REVIEW

**Comment by nimid** [comment]
@guzzilar 

> Add condition, if store's currency is not one of "THB", "JPY", "IDR", "SGD". Omise card form will be hidden.

Although, to hide the Omise card form can helps to stop the un-preferred result from create charge with un-support currency but to hide the Omise card form can make the confusion that what is the reason that it is not display. The cause that result to Omise card from not display can come from another reason such as programming error, etc.

So, it is not directly indicate for the merchant and developer.

How about the alternate solutions?

1. Display the information message. For an example `Checkout with Credit card / Debit card is not support for the current currency`.

2. Display the Omise card form normally but display the error that responded from the Omise API. Below is the response from Omise API when create charge with un-support currency.

```json
{
  "object": "error",
  "location": "https://www.omise.co/api-errors#invalid-charge",
  "code": "invalid_charge",
  "message": "currency is currently not supported"
}
```

**Comment by guzzilar** [comment]
rebased PR #40 (`1-code-styling`) to PR #44 (`1-dev-currency-support`)

**Comment by guzzilar** [comment]
@nimid I agree with your thought.

But for the alternate solutions, Im ok with the first one (the second one would make buyer has to do `choose Omise payment method`, `input their card information`, `submit payment` to get the information that they cannot make a charge with credit card method.)

Actually I thought that the perfect solution is something like a screenshot below (disable the option & show warning message),
but can't make it on Magento.
screen shot 2560-02-01 at 1 26 52 am

Btw, if your concern is about merchant and developer side, I would add message or something at the admin page instead, to inform merchant and developer that their current currency doesn't support for Omise payment method and the payment method from the checkout page will be disappeared.

And keep hiding the Omise payment method from the checkout form as it is (for buyer side that they will not be confused with the payment options)

**Comment by guzzilar** [comment]
rebased with `1-stable`

**Comment by guzzilar** [comment]
@nimid as for our conversation, 
- [pull/44#issuecomment-275616940](https://github.com/omise/omise-magento/pull/44#issuecomment-275616940)

- [pull/44#issuecomment-276458059](https://github.com/omise/omise-magento/pull/44#issuecomment-276458059)

Is it possible to move this topic to another PR (or even can [create new issue](https://github.com/omise/omise-magento/issues)) ?
Since the conversation tended to be an enhancement topic.

**Comment by nimid** [comment]
@guzzilar I found another solution to display a message when the current currency is not supported.

Can you look at the pull request #50 ?

The pull request #50 is for example only. It will be delete after discussed.

**Comment by nimid** [comment]
@guzzilar The screenshot below shows the example at the front-end checkout form, when the payer proceed the checkout with the current currency is not supported.

example_message

**Comment by guzzilar** [comment]
@nimid checked, LGTM. 👍

**Comment by guzzilar** [comment]
@nimid do you want to merge that PR to here? or better to just push new comment within this branch.

**Comment by guzzilar** [comment]
*I meant, push new commit

**Comment by nimid** [comment]
@guzzilar The change in the pull request #50 is not completed. At least 2 parts, I found it need to be improved.

1. If the message will contain the payment name (`Credit Card (Powered by Omise)`), the source code must not be static, it should be retrieved from the configuration
2. The indentation in `omisecc.phtml` at the part of new condition should be corrected

And it has no any description. So, process or work on the normal way is easier.

For the pull request #50, I would like it to be discussed. It can show the working change that already proved that the concept, idea, or suggestion is surely worked.

I found this way is better when compare with only show some part of code, dry sample or the example that not actually worked.

Another way that I found it is good is notify the merchant instantly when they change the currency to be the one that Omise did not supported. But this way is harder and more expensive. It need to know more about the core source code of Magento and may be it need to change or modify the core source code of Magento.

**Comment by guzzilar** [comment]
@nimid for 
1. I think it's fine for now, since we don't have the setting that let merchant change payment title by themselves.

2. 👌 

So, would you like to push new commit by yourself or want me to do it?

**Comment by guzzilar** [comment]
@nimid just commented at https://github.com/omise/omise-magento/pull/50#issuecomment-278039915

File: src/app/code/community/Omise/Gateway/Model/PaymentMethod.php [hunk 1]
```diff
@@ -135,4 +135,47 @@ public function assignData($data)
...
+    {
+        switch (strtoupper($currency)) {
+            case 'THB':
+            case 'IDR':
+            case 'SGD':
+                // Convert to a small unit
+                $amount = $amount * 100;
+                break;
+
+            case 'JPY':
```
Comment (nimid) [inline src/app/code/community/Omise/Gateway/Model/PaymentMethod.php]: @guzzilar If the currency does not require formatting, I think this condition can be removed.

For example, if another currency has been opened to receive payment and it does not require formatting. To comply with this pattern, it need to add another empty condition.

For an example with Philippine peso,

```php
case 'JPY':
    break;

case 'PHP':
    break;
```

And maybe more currencies.

**Comment by nimid** [comment]
@guzzilar 

I have read a comment, https://github.com/omise/omise-magento/pull/50#issuecomment-278039915.

> So, would you like to push new commit by yourself or want me to do it?

Now, it has another expanded approach from a comment, https://github.com/omise/omise-magento/pull/50#issuecomment-278039915.

So, it up to you to do it.

For me, I am going to delete the pull request #50.

File: src/app/code/community/Omise/Gateway/Model/PaymentMethod.php (hunk 1 above)
Comment (guzzilar) [inline src/app/code/community/Omise/Gateway/Model/PaymentMethod.php]: 👌 will remove it.

**Comment by guzzilar** [comment]
@nimid  https://github.com/omise/omise-magento/pull/44#issuecomment-278048436
I'll implement it now.

Thanks for suggestion. 👍

**Comment by guzzilar** [comment]
@nimid 
done here 
- https://github.com/omise/omise-magento/pull/44/commits/2384086b4ef6bba4c1e222a78c19b407df226618
- https://github.com/omise/omise-magento/pull/44/commits/437635043c8153451dcb65c2d12010686f626c80

And updated test number 5 at the section 3.

**Comment by nimid** [comment]
@guzzilar The first bullet at the section, 2. Description of change, requires change to be conformed with the current update.

**Comment by guzzilar** [comment]
@nimid nice catch 👍 , updated

**Comment by nimid** [comment]
@guzzilar At the section, 3. Quality assurance, the detail 10th, the amount `$54,000` should be added a character `S` to make it conform with the actual result and screenshot.

**Comment by guzzilar** [comment]
@nimid done, nice catch again 👍

**Comment by nimid** [comment]
@guzzilar This pull request is ready to merge.

The conclusion below is about the new support for a currency SGD which it has not been available at the Omise API side.

- When the currency SGD is available at the Omise API side, it still require the actual system test. Although, this pull request has provided the result of test on the 4th and 10th but the 4th is the result at the error log on Omise dashboard and the 10th is test by hard-code.

For this situation, another test strategy such as unit test or another test can help.

**Comment by guzzilar** [comment]
☝️ added to PR description, 6th additional notes.

👍

**Pull request merged**